(WIP) Fix #1229: add managed dependency version (29.0-jre) for Guava

tatu-at-datastax opened the pull request:
**What this PR does**:

Adds managed dependency version `29.0-jre` for Guava, removes explicit version definitions from poms

**Which issue(s) this PR fixes**:

Fixes #1299

**Checklist**
- [n/a] Changes manually tested -- dependency upgrade tested via existing unit, integ, func tests
- [n/a] Automated Tests added/updated -- same as above
- [n/a] Documentation added/updated -- should not change visible behavior.
-

tatu-at-datastax commented:
Ready for review (CI passed).

ivansenic reviewed (changes requested):
It would really be great if we could fix now also the apparently wrong imports of the shadowed guava.. I think every non-persistence module should use guava directly.. If the dependency is not in the pom explicitly we should add it..

Futhermore, maybe we can quickly investigate how can we ban this in future. The shadowed guava should not be on the compile path for those modules.. I have seen other deps to be also available in the compile, but then at run-time not.. I am pritty sure I hit that with `MutableInt (Commons Lang 2.6 API)`, but when using same from `lang3` it works..

@tatu-at-datastax adding request changes, but we can also split in another PR if you like it better..

ivansenic commented:
@tatu-at-datastax also see Dependabot alert about the Guava version..

tatu-at-datastax commented:
@ivansenic Ah yes. I was curious about shadowed guava and +1 for removing that if possible. I think I'd like to do that in a separate issue/PR.

Also agree with explicit dependency: I'll see if I can find usages, add those explicitly.
Guava as a dependency can be pretty gnarly.

tatu-at-datastax commented:
Oh and Dependabot. Yeah, maybe it makes sense to boost default version to even higher -- if I recall, that CVE is not often actually relevant, but all security tools flag it so it's just easier to upgrade. We should be able to do that.

ivansenic reviewed (approved):
Changing to approved as we agreed to move suggestions to other PRs...

tatu-at-datastax commented:
@ivansenic @dougwettlaufer I will do some of simple changes here still, moving back to WIP. Apologies for confusion. Will let you know once this is ready

tatu-at-datastax commented on config-store-yaml/pom.xml:
```diff
@@ -41,7 +35,7 @@
         <groupId>com.fasterxml.jackson.dataformat</groupId>
         <artifactId>jackson-dataformat-yaml</artifactId>
-        <version>2.11.3</version>
+        <version>2.12.5</version>
```
In near future, I'll probably try to add `dependencyManagement` for Jackson (similar to Guava here); challenges are similar as Stargate can use newer/-est versions just fine but C*3.11/4.0/DSE depend on older versions (although I am working on trying to get Jackson update for Cassandra too, concurrently).

tatu-at-datastax commented on grpc/pom.xml:
```diff
@@ -52,6 +52,10 @@
       <version>${grpc.version}</version>
     </dependency>
     <!-- Other -->
+    <dependency>
```
Has direct usage, adding dependency.

ivansenic replied:
Here is the same, some gRPC dep was introducing guava and we now overwrote the version most likely

tatu-at-datastax replied:
Changed the approach to avoid this issue.

tatu-at-datastax commented on health-checker/pom.xml:
```diff
@@ -101,11 +95,6 @@
       <groupId>io.dropwizard</groupId>
       <artifactId>dropwizard-logging</artifactId>
     </dependency>
-    <dependency>
```
Removed minimal usage (see below), and hence dependency and exclusions as well.

tatu-at-datastax commented on persistence-test/pom.xml:
```diff
@@ -52,9 +52,12 @@
       <scope>compile</scope>
     </dependency>
     <dependency>
+      <!-- 09-Sep-2021, tatu: Cassandra 3.11/4.0 backends depend on Jackson 2.9;
```
Could probably have worked around this by forcing `jackson-core` and `jackson-databind` versions to 2.12.5, but due to compatibility concerns with backends decided to limit scope here, only boosting to 2.9.10 that C*/DSE use.

tatu-at-datastax commented on pom.xml:
```diff
@@ -38,6 +38,7 @@
     <commons-io.version>2.7</commons-io.version>
     <immutables.version>2.5.6</immutables.version>
     <grpc.version>1.35.0</grpc.version>
+    <guava.version>30.0-jre</guava.version>
```
As per Dependabot warning about Guava version <= 29.0

tatu-at-datastax commented:
@ivansenic @dougwettlaufer I did bit more work as suggested, added notes on changes -- please let me know if there's anything else to address here. I will create a new issue for removing/reducing use of shaded guava dependency for follow-up work (and maybe one for Jackson deps too)

ivansenic commented on cql/pom.xml:
```diff
@@ -63,6 +63,10 @@
       <artifactId>caffeine</artifactId>
       <version>2.8.5</version>
     </dependency>
+    <dependency>
```
OK wait, but are you sure there is no transitive deps here of guava, meaning that something else was using Guava but now we override the version? I was thinking to add dep only where shadowed Guava was used and we wanted to move away from it.. Let's sync on this quickly..

tatu-at-datastax replied:
Changed inclusion to shaded version.

ivansenic commented on health-checker/src/main/java/io/stargate/health/CheckerResource.java:
```diff
@@ -40,7 +39,7 @@ public CheckerResource() {
   }
 
-  @VisibleForTesting
+  //  @VisibleForTesting // the only use of Guava from this module
```
Delete the line?

tatu-at-datastax replied:
> Delete the line?

Sure, will do.

ivansenic reviewed (changes requested):
I have some worries about this.. Let's sync please..

tatu-at-datastax commented:
@ivansenic I added explicit Guava dependency in cases where module itself depends on guava (grepped for imports), removed if not. So did not leave dependency for transitive usage.

ivansenic commented:
@tatu-at-datastax What's the state with this and what path are we taking at the end.. Maybe we can talk shortly today?

tatu-at-datastax commented:
@ivansenic yes, was about to add a note. Let's sync up today.

My plan was to remove usage of un-shaded Guava, along with reducing usage in general where applicable.

tatu-at-datastax commented:
Closing this in favor of #1253 (although going through similar issues over there -- but finally making progress).
